core/state/statedb.go

```diff
@@ -838,11 +847,21 @@ func (s *StateDB) IntermediateRoot(deleteEmptyObjects bool) common.Hash {
 	// If there was a trie prefetcher operating, terminate it async so that the
 	// individual storage tries can be updated as soon as the disk load finishes.
 	if s.prefetcher != nil {
-		s.prefetcher.terminate(true)
-		defer func() {
+		if s.collectWitness {
+			// terminate prefetcher synchronously to force resolution of all pending
+			// fetches so that nothing is missing from the witness.
+			s.prefetcher.terminate(false)
 			s.prefetcher.report()
-			s.prefetcher = nil // Pre-byzantium, unset any used up prefetcher
-		}()
+			defer func() {
+				s.prefetcher = nil
+			}()
+		} else {
+			s.prefetcher.terminate(true)
+			defer func() {
+				s.prefetcher.report()
+				s.prefetcher = nil // Pre-byzantium, unset any used up prefetcher
+			}()
+		}
```

Comment:
This was completely meaningless. The async prefetcher just returns, but will block when attempting to retriebve the trie. The whole thing is synchronous already in the scope of the method. Please revert.

Reply:
I.e. your code blocks everything until everything is done, **and then** starts the storage trie updates. The original code just signals everything to stop and **concurrently** waits for each and does the storage trie updates as soon as the individual prefetchers terminate, running concurrently instead of waiting for all and then messing with the updates. This is a 5-7% speed update you nuked out here.